### PR TITLE
fix: MXFP4 mantissa rounding

### LINKED ISSUE #974: [Issue]: Incorrect MXFP4 mantissa rounding
#### Issue with normal values

We have an FP32 tensor defined as follows:
```
torch.tensor([
    0.3463, -0.0499, 0.3794, -2.3764, 0.3960, 0.5989, -0.6250, -0.4239,
    -0.6192, 0.2761, -0.3746, 0.1462, 0.5216, -0.5194, -0.8497, -0.8237,
    0.5145, 0.3508, 0.0298, -0.3462, 0.3917, 0.1096, 1.4494, 1.7009,
    -0.5473, -2.0851, 1.8099, 0.3587, 1.4622, -2.3143, 0.3412, -0.8931
])
```

With a UE8M0 scale of 126, the packed FP4 generated by [`_mxfp4_quant_op`](https://github.com/ROCm/aiter/blob/6b586aee05672f8e2fc2d2d9103a124ded13bd78/aiter/ops/triton/quant.py#L197)  is

```
[129, 226,  34, 171,  26,  25, 162, 187,  18, 144,   2,  85, 234,  22, 229, 193]
``` 

However, torchao has a different result which is aligned with MI355 instruction `v_cvt_scalef32_pk_fp4_f32` :

```
[129, 226,  34, 170,  26,  25, 162, 187,  18, 144,   2,  85, 234,  22, 229, 193]
```

Specifically, the value `-0.6250 / 0.5` is rounded to -1 instead of -1.5.

#### Issue with denormal values

Denormal values cannot be rounded up in the current aiter implementation.

For example:

FP32 numbers: 0x3F000000 and 0x3F000003
ue8m0 scale: 128

torchao and `v_cvt_scalef32_pk_fp4_f32` round 0x3F000000 to 0 and 0x3F000003 to 0.5, 
while the current aiter rounds both to 0.5.

### Operating System

Ubuntu 22.04

### CPU

AMD EPYC 73F3 16-Core Processor

### GPU

AMD Instinct MI355X

### ROCm Version

ROCm 7.0

### ROCm Component

_No response_

### Steps to Reproduce

_No response_

### Additional Information

Possible fix:

https://github.com/AMD-AGI/torchtitan-fa-fp8/blob/han/mxfp4_refactor/torchtitan/experiments/kernels/blockwise_fp4/mxfp_quantization.py#L50-L121

### PR DESCRIPTION
**Pull request by hann-wang**
fixes #974

### CODE REVIEW

**Comment by carlushuang** [comment]
cc @vgokhale

File: aiter/ops/triton/_triton_kernels/quant.py [hunk 1]
```diff
@@ -125,26 +135,49 @@ def _mxfp4_quant_op(
...
+    normal_x = normal_x >> (MBITS_F32 - MBITS_FP4)
+    normal_x = normal_x.to(tl.uint8)
+
+    # Merge results
+    e2m1_value = tl.full(qx.type.get_block_shapes(), 0x7, dtype=tl.uint8)
+    e2m1_value = tl.where(normal_mask, normal_x, e2m1_value)
+    e2m1_value = tl.where(denormal_mask, denormal_x, e2m1_value)
+    # add sign back
+    sign_lp = s >> (MBITS_F32 + EBITS_F32 - MBITS_FP4 - EBITS_FP4)
+    sign_lp = sign_lp.to(tl.uint8)
```
[the remark this answers is not in the document]
Comment (hann-wang) [reply aiter/ops/triton/_triton_kernels/quant.py:179]: The triton kernel uses UINT32 data type and hence no need of the sign mask. PyTorch does not have bit shift operations for UINT32, and we are using INT32 instead. That's where the sign mask comes in.

File: aiter/ops/triton/_triton_kernels/quant.py [hunk 2]
```diff
@@ -125,26 +135,49 @@ def _mxfp4_quant_op(
...
+    # rounding bias part 2
+    normal_x += mant_odd
+    # take the bits!
+    normal_x = normal_x >> (MBITS_F32 - MBITS_FP4)
+    normal_x = normal_x.to(tl.uint8)
+
+    # Merge results
+    e2m1_value = tl.full(qx.type.get_block_shapes(), 0x7, dtype=tl.uint8)
+    e2m1_value = tl.where(normal_mask, normal_x, e2m1_value)
+    e2m1_value = tl.where(denormal_mask, denormal_x, e2m1_value)
```
[the remark this answers is not in the document]
Comment (hann-wang) [reply aiter/ops/triton/_triton_kernels/quant.py:176]: updated the unit test instead.

**Comment by hann-wang** [comment]
> Hi @hann-wang, can you please do a merge/rebase from main? Looks like there are some conflicts that need to be solved before merging your PR.

Hi @lucas-santos-amd , I just merged the changes from main.

**Review by lucas-santos-amd — APPROVED** [review]
